pkg/sentry/fsimpl/devpts/line_discipline.go

```diff
@@ -455,6 +456,80 @@ func (*inputQueueTransformer) transform(l *lineDiscipline, q *queue, buf []byte)
 			l.terminal.replicaKTTY.SignalForegroundProcessGroup(kernel.SignalInfoPriv(linux.SIGQUIT))
 		}
 
+		// In canonical mode, some characters need to be handled specially; for example, backspace.
+		if l.termios.LEnabled(linux.ICANON) {
+			// n_tty.c:n_tty_receive_char_canon
+			c := cBytes[0]
+			c_cc := l.termios.ControlCharacters
+
+			// c == c_cc[linux.VKILL] is also handled by n_tty.c:eraser, but this isn't implemented
+			if c == c_cc[linux.VERASE] || (c == c_cc[linux.VWERASE] && l.termios.LEnabled(linux.IEXTEN)) {
+				// n_tty.c:eraser
+				var kill_type int
+				if c == c_cc[linux.VERASE] {
+					kill_type = linux.VERASE
+				} else {
+					kill_type = linux.VWERASE
+				}
+				seen_alnums := 0
+				for len(q.readBuf) > 0 {
+					// Erase a single possibly multibyte character
```

Comment:
I deliberately tried to follow the Linux kernel's source code as closely as possible to reduce mistakes and make it easier to port over more features later, but I'm happy to rewrite this.

Just curious, are there copyright restrictions that prevent us from using code verbatim? Why do we prefer to avoid reusing code and comments from the Linux kernel?

Thanks for reviewing this PR and for your detailed feedback!

Reply:
Generally speaking, I believe GPLv2-licensed code cannot be used in an Apache-licensed project.

But regardless of licensing, gVisor does not aim to be a copy of Linux. The goal is to be a secure sandbox, and part of how it achieves this is by being a reimplementation from the ground up. In general, that means looking at man pages and documentation of how the syscall ABI should work, and implementing Go code from that specification, rather than from the Linux implementation. Doing it this way decreases the probability of duplicating potential vulnerabilities in Linux into gVisor, relative to what would happen if code was merely ported over.

There are parts of Linux behavior which aren't fully covered by man pages and documentation, but need to be faithfully emulated in gVisor to ensure compatibility of Linux applications. In those cases, gVisor does try to reproduce Linux's behavior. But the way we enforce this consistency is by black-box tests (`pty_test.cc`) which verify the similarity between Linux and gVisor behavior by actually invoking those syscalls and seeing what happens, rather than by copying code over.

Reply:
I see. Thank you for explaining this to me!